### DISCUSSION TITLE
Issue with ServiceStack 8.2.2 and .NET Framework 4.8

### DISCUSSION THREAD

**Discussion by cscawley** [ServiceStack/Discuss#119] [Q&A]
I have an issue upgrading a VB .NET application from ServiceStack 3.X to 8.2.2

I'm using an InMemoryAuthRepository and injecting ICacheClient into our custom credential auth repository which has overrides for TryAuthenticateAsync where credentials are checked using Microsoft Identity after success of which a session is created. All this works except after the TryAuthenticateAsync override completes, the callstack ends up in the RedisAuthRepositoryAsync class somehow and fails because there is no Redis connection. That class is expecting a connection object to be instantiated. Declaratively I'm explicitly using InMemory and nothing in my code is referencing or declaring Redis.

Also I used the latest VB ServiceStack scaffolding to get a quick proof of concept in .NET 6 with the same InMemory/ICacheClient declarations and TryAuthenticateAsync override and it works successfully without ever traversing that Redis class.

Does anyone have any insight on using the latest ServiceStack with .NET Framework?

```vb
Public Class Global_asax
    Inherits HttpApplication

    Public Class ServiceAppHost
        Inherits AppHostBase
          'other app initialization
          Public Overrides Sub Configure(container As Funq.Container)
            'other configure code
                Dim userRep = New InMemoryAuthRepository()
                container.Register(Of IUserAuthRepository)(userRep)
                container.AddSingleton(Of ICacheClient)(New MemoryCacheClient())
                Dim cacheClient = container.Resolve(Of ICacheClient)() ' Resolve the cachedClient instance
                Dim af As AuthFeature = New AuthFeature(Function() New AuthUserSession(), New IAuthProvider() {New CognitoAuthProvider(cacheClient)}, Nothing)
                af.IncludeAssignRoleServices = False
                Plugins.Add(af)
            End Sub
    End Class
    Sub Application_Start(ByVal sender As Object, ByVal e As EventArgs)
        Dim appHost As New ServiceAppHost
        Licensing.RegisterLicense(ConfigurationManager.AppSettings("servicestack:license"))
        appHost.Init()
    End Sub
End Class
```

**Comment by mythz** [comment]
Can you provide the StackTrace for the Redis Exception

**Reply by cscawley** [reply]
Absolutely and thanks for the reply.
Here's the stack trace. I've also added 7 threads of the call stack that precede it:

```
System.NullReferenceException
  HResult=0x80004003
  Message=Object reference not set to an instance of an object.
  Source=ServiceStack
  StackTrace:
   at ServiceStack.Auth.RedisAuthRepository`2.<GetUserAuthAsync>d__64.MoveNext() in /home/runner/work/ServiceStack/ServiceStack/ServiceStack/src/ServiceStack/Auth/RedisAuthRepositoryAsync.cs:line 285

ServiceStack.dll!ServiceStack.Auth.RedisAuthRepository<ServiceStack.Auth.UserAuth, ServiceStack.Auth.UserAuthDetails>.GetUserAuthAsync(ServiceStack.Auth.IAuthSession authSession, ServiceStack.Auth.IAuthTokens tokens, System.Threading.CancellationToken token) Line 277	C#

ServiceStack.dll!ServiceStack.Auth.AuthProvider.ValidateAccountAsync(ServiceStack.IServiceBase authService, ServiceStack.Auth.IAuthRepositoryAsync authRepo, ServiceStack.Auth.IAuthSession session, ServiceStack.Auth.IAuthTokens tokens, System.Threading.CancellationToken token) Line 414	C#

ServiceStack.dll!ServiceStack.Auth.AuthProvider.OnAuthenticatedAsync(ServiceStack.IServiceBase authService, ServiceStack.Auth.IAuthSession session, ServiceStack.Auth.IAuthTokens tokens, System.Collections.Generic.Dictionary<string, string> authInfo, System.Threading.CancellationToken token) Line 219	C#

ServiceStack.dll!ServiceStack.Auth.CredentialsAuthProvider.AuthenticateAsync(ServiceStack.IServiceBase authService, ServiceStack.Auth.IAuthSession session, string userName, string password, string referrerUrl, System.Threading.CancellationToken token) Line 164	C#

ServiceStack.dll!ServiceStack.Auth.CredentialsAuthProvider.AuthenticateAsync(ServiceStack.IServiceBase authService, ServiceStack.Auth.IAuthSession session, ServiceStack.Authenticate request, System.Threading.CancellationToken token) Line 126	C#

ServiceStack.dll!ServiceStack.Auth.AuthenticateService.AuthenticateAsync(ServiceStack.Authenticate request, string provider, ServiceStack.Auth.IAuthSession session, ServiceStack.Auth.IAuthProvider oAuthConfig, System.Threading.CancellationToken token) Line 517	C#

ServiceStack.dll!ServiceStack.Auth.AuthenticateService.PostAsync(ServiceStack.Authenticate request) Line 293	C#
```

**Reply by mythz** [reply]
The issue is because of a `NullReferenceException`, it's not attempting to make a Redis connection, the Memory and Redis Auth Providers share the same implementation since they're implemented in the same way but the `IRedisClientFacadeAsync` instance in the `InMemoryAuthRepository` uses an [InMemoryManagerFacade](https://github.com/ServiceStack/ServiceStack/blob/fbca1cfe4916091b783464beb941881e6f78d264/ServiceStack/src/ServiceStack/Auth/InMemoryAuthRepository.cs#L61) which handles all calls in memory.

Although it's not clear why you would get an NRE on that line:

https://github.com/ServiceStack/ServiceStack/blob/main/ServiceStack/src/ServiceStack/Auth/RedisAuthRepositoryAsync.cs#L285

Since the `authSession` is the only object that could throw an NRE is guarded by an if statement. 

I'd recommend debugging into that class and putting a breakpoint on that line to find out what's causing the NRE.

**Reply by cscawley** [reply]
So, an interesting thing happens when we debug the call right before the NRE. The authSession object is there and a user is authenticated, but the IAuthTokens object is null. When the call stack enters the non-virtual GetUserAuthAsync it gets past that conditional statement with IAuthSession instantiated but the exception still isn't very helpful.

<img width="777" alt="IAuthToken_null" src="https://github.com/ServiceStack/Discuss/assets/5122906/30990dde-9b4f-4f78-ac56-55216f9d58bb">

<img width="819" alt="IAuthSession" src="https://github.com/ServiceStack/Discuss/assets/5122906/76001989-c8f8-4eaa-b43a-a0e64412b5cf">

<img width="764" alt="IAuthSession_IAuthTokens" src="https://github.com/ServiceStack/Discuss/assets/5122906/33edc098-ed6c-4531-ac25-d47b88ce45ff">

**Reply by mythz** [reply]
`tokens` being null shouldn't cause the NRE, can you put a breakpoint inside the method its calling.

**Reply by cscawley** [reply]
Ok I see. Well it looks like on this GetUserAuthAsync member it's doing a TryParseInt64 on the userAuthId. We've assigned the Microsoft Membership user ID to this field which is a GUID. That GUID is failing to parse to int64 and it's returning null. Simply by changing that assignment to any random integer fixes everything.
![Screenshot 2024-04-11 065159](https://github.com/ServiceStack/Discuss/assets/5122906/fcd79809-a282-4b56-8f81-48203ebaa15c)
